Loading classes from external Maven project

jdiazgon opened the pull request:
This pull request contains all the needed changes for loading classes from external Maven project. This is very useful for our next CobiGen CLI.

### Implements:

- Implemented Class loader for external Maven project. Tries first to load classes from `eclipse-target` on both the `core` and the `api` module. If not found, tries to load classes from `target`.
- Recursive way to load dependencies from depending POMs. For instance, If my current project is defining as dependency `devon4j-rest`, we don't only need to load the classes from `devon4j-rest` but also the ones that are defined as dependency, like `devon4j-service`.
This recursiveness is limited up to 4 levels, because if not we will have performance issues.

@hohwille what should be done with the formatting differences? I tried to use devonfw formatting but it is different to yours.

jdiazgon commented:
Only my test is failing `testContextLoadingFromLocalMavenProject(...)`. I thought that it could be because in test resources I have a Maven project that is needed for the test and maybe I had to do a `mvn install` there.

Unfortunately, it seems that didn't fix it.

hohwille commented on .gitignore:
```diff
@@ -6,5 +6,11 @@
+!/java/impl/src/test/resources/testdata/localmavenproject/maven.project/core/target
+!/java/impl/src/test/resources/testdata/localmavenproject/maven.project/core/eclipse-target
+!/java/impl/src/test/resources/testdata/localmavenproject/maven.project/api/target
+!/java/impl/src/test/resources/testdata/localmavenproject/maven.project/api/eclipse-target
```
Adding binary class files to git is IMHO not an acceptable approach.
Please remove this lines and the added *.class files
Instead introduce a maven build via module (or if required via exec plugin) to mak tests work.

jdiazgon commented:
So, I have tried to run the tests on a Centos 7 machine, and every test is passing:

![image](https://user-images.githubusercontent.com/31688036/58568959-7b98cd80-8235-11e9-8774-9b62c2edefb9.png)

@hohwille do you have any idea why is this failing on Travis? I'm just getting a headache with this.

hohwille commented:
@jdiazgon Thanks for your improvements. Excellent that you solved the maven.project issue and removed the binary content from git.
Can you still reformat the code after fixing your code formatter so I can properly read the diffs?

hohwille commented:
Travis is indeed strange here:
```
[ERROR] testContextLoadingFromLocalMavenProject(net.sf.mmm.code.impl.java.JavaExtendedContextWithMavenAndSourceCodeTest)  Time elapsed: 0.023 s  <<< FAILURE!
java.lang.AssertionError
	at net.sf.mmm.code.impl.java.JavaExtendedContextWithMavenAndSourceCodeTest.getContextFromLocalMavenProject(JavaExtendedContextWithMavenAndSourceCodeTest.java:54)
	at net.sf.mmm.code.impl.java.JavaExtendedContextWithMavenAndSourceCodeTest.testContextLoadingFromLocalMavenProject(JavaExtendedContextWithMavenAndSourceCodeTest.java:98)
```
According to your code:
https://github.com/jdiazgon/code/blob/externalMaven/java/impl/src/test/java/net/sf/mmm/code/impl/java/JavaExtendedContextWithMavenAndSourceCodeTest.java#L54

There is no assertion in that line. So IMHO this is an incomplete stacktrace. The important information is in the `target/surefire-report` directory on the travis docker machine that is destroyed after the build so we can not access further details for debugging.

hohwille commented:
OK. Test is failing reproducable in this assertion:
https://github.com/jdiazgon/code/blob/externalMaven/base/src/main/java/net/sf/mmm/code/base/loader/BaseSourceCodeProviderArchive.java#L49

hohwille commented:
archiveFile is `.../code/java/impl/src/test/resources/testdata/localmavenproject/maven.project/core/src/test/java` so this is a non-existing folder. On your machine the folder exists locally but as it has no content it does not exist in git and therefore on other machines that cloned this git state.

hohwille commented:
This is a bug from myself in JavaSourceProviderUsingMaven
```
    private BaseSourceLoader createLoader(File sourceCodeLocation) {

        SourceCodeProvider sourceCodeProvider;
        if (sourceCodeLocation == null) {
            sourceCodeProvider = null;
        } else if (sourceCodeLocation.isDirectory()) {
            sourceCodeProvider = new BaseSourceCodeProviderDirectory(sourceCodeLocation);
        } else {
            sourceCodeProvider = new BaseSourceCodeProviderArchive(sourceCodeLocation);
        }
        return new JavaSourceLoader(sourceCodeProvider);
    }
```
It does not check if the file exists at all and assumes that if it is no directory, it must be an archive (JAR) but this assumption is wrong if the file does not exist.

hohwille commented:
We simply need to add a check here if the file exists:
```
    if ((sourceCodeLocation == null) || !sourceCodeLocation.exists()) {
```

Can you fix this?

hohwille commented:
However, I still get a NPE if I fix that locally because the maven model is null.
I debugged through your `getApiByteCodeLocations` and have to say that this is not a solution. You are coding wired assumptions here that might work for a devon4j project but will not work in general. Instead we need to parse the POM hierarchy to find parent modules properly. Searching and guessing for an `api` folder is just a quick hack. Unfortunately I am very busy and overload in the next 3-4 weeks. Otherwise I could take over that part. So you have to wait or get trough all this without me.

hohwille commented:
You could even subscribe to maven developer list and ask if there is an API for recursively reading parent POMs as long as they come from the local disc and collect all modules building the project reactor of if that all has to be written by hand
https://maven.apache.org/mailing-lists.html

jdiazgon commented:
> Can you still reformat the code after fixing your code formatter so I can properly read the diffs?

Actually I tried to use two formatter:

1. I used CobiGen-IDE formatter, but it is different from yours.
2. I used devonfw-IDE formatter, but it is different from yours.

I think that you are using a personalized formatter configuration, which I don't have. Could you send me your configuration?

jdiazgon commented:
> Can you fix this?

Yes! I will try. Thanks for the debugging :)

> I debugged through your getApiByteCodeLocations and have to say that this is not a solution

I agree, I will try to find a better approach.

jdiazgon commented:
> However, I still get a NPE if I fix that locally because the maven model is null.

I found out why you were still getting an NPE. The problem is that your `maven.project` project test didn't have the compiled classes (the `target` folder was not present). This is because you did `mvn clean install` inside the `impl` project.

The issue is the following:

- In order to install the `maven.project` project test I added it as a module on `mmm-code-java`, please take a look [here](https://github.com/m-m-m/code/pull/18/commits/9b875342df5019425458ae9b67a547ce87f2ebff#diff-61657881d138c2b7ac1366d88c9ee035R20).

  This means that you need to execute `mvn clean install` on `mmm-code-java` rather than on `mmm-code-java/impl` project. This is a bit dirty, so I tried to add the module on the `impl` pom. The problem with that pom is that it defines this `<packaging>jar</packaging>`. So when I tried to add the module, Maven said:

```java
[ERROR] 'packaging' with value 'jar' is invalid. Aggregator projects require 'pom' as packaging
```

What should we do here? Currently, I'm going to try the exec plugin approach.

jdiazgon commented:
Now that seems fixed :) I have used the exec plugin.

hohwille commented:
A `<module>` declaration in a maven POM is just a path. It can be a single folder, but may also be multiple folders including stuff like `../foo/./nonsense`. So any relative path should work.

hohwille commented:
> I think that you are using a personalized formatter configuration, which I don't have. Could you send me your configuration?

There might be nuonces that differ because we seem to have different versions of Eclipse. But what I am complaining about is that you are using 4 spaces as indendation and devonfw is using 2 as a standard. I also use 2 spaces in mmm. Look at any code in mmm or devonfw (except CobiGen that indeed is using 4 spaces and does not follow the devonfw standard) - examples:
* https://github.com/devonfw/devon4j/blob/develop/modules/basic/src/main/java/com/devonfw/module/basic/common/api/to/AbstractTo.java
* https://github.com/devonfw/sonar-devon-plugin/blob/master/src/main/java/com/devonfw/ide/sonarqube/common/api/JavaType.java
* https://github.com/m-m-m/code/blob/master/api/src/main/java/net/sf/mmm/code/api/CodeContext.java

The indendation size for devonfw is 2. If you do not believe me look here:
* https://github.com/devonfw/devon-ide/blob/master/settings/src/main/settings/eclipse/workspace/update/.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.jdt.core.prefs#L121
* https://github.com/oasp/oasp4j-ide/blob/master/oasp4j-ide-settings/src/main/settings/eclipse/workspace/update/.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.jdt.core.prefs#L121
* https://github.com/devonfw/devon4j/blob/develop/.editorconfig#L4

jdiazgon commented:
I think I have already sent a e-mail to the dev mailing list, but meanwhile we wait I did the following:

The last two commits I think that already fix the problem  of reading parent POMs. The method that implements this logic is [this one](https://github.com/m-m-m/code/blob/0bf2da2eba976039dc96dffa096e0e05d26ed73a/java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java#L286). We already had an API to load parent POMs from, and also we are able to read which `modules` does it contain.

Now I will focus on fixing the formatting issues.

jdiazgon commented:
As you suggested, I have used two spaces indentation and now it seems much better. From my side the code is ready to be merged.

hohwille commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java:
```diff
@@ -129,71 +140,276 @@ private BaseSource createSource(Dependency dependency, File byteCodeArtifact) {
+        Model parentModel = parseModel(parentPom);
+
+        if (parentModel == null) {
+          return dependenciesURLs;
```
better `break` to go the normal exit via return at the end of the method. Some day someone might need to add a post-processing to the result before returning. The he may miss side-exists.

hohwille commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java:
```diff
@@ -129,71 +140,276 @@ private BaseSource createSource(Dependency dependency, File byteCodeArtifact) {
+              moduleLocation = moduleLocation + segment + "/";
+            }
+            i++;
+          }
```
do not fiddle with path via such string operations. Instead simply use Java File or Path API.

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java:
```diff
@@ -30,170 +31,385 @@
+     */
+    private ArrayList<URL> getDependenciesURLS(Model model, ArrayList<URL> previousURLs, int recursiveness) {
+
+        if (recursiveness >= 4) {
```
I agree that it would be very nice to have a lazily loading of the dependencies. However, this seems a very complex topic. For instance:

When I try to load `EmployeeEntity`, I first need to load `Employee` defined on the `api`. This is easy because the `core` already defined `api` as a dependency. But, `Employee` extends `ApplicationEntity` which extends `GenericEntity`. That class is tricky to load because it is not defined inside the `api`.

The `api` defines as dependency `devon4j-rest` which defines as dependency `devon4j-basic`. So it is indirectly referenced.

If Maven was created, was in order to fix this [JAR hell](https://dzone.com/articles/what-is-jar-hell) problem.  If you run the command `mvn dependency:build-classpath` in any of your Maven projects, you will see all the jars that are loaded in the class path.

If you guide me on this, we can try to find an optimal solution. However, my proposed solution already works pretty well. We have even tested using My Thai Star, which as you know, is a large project.

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/JavaContext.java:
```diff
@@ -195,15 +197,19 @@ public BaseContext getContext() {
         LOG.debug("Class {} not found.", qualifiedName, e);
         return null;
+      } catch (NoClassDefFoundError e) {
+        LOG.debug("Failed to find a dependency.", e);
```
> getType("none.existing.Type"). So LOG.debug("Could not find class {}.", qualifiedName); would make more sense.

I agree, I have changed it.

> but must not be a failing dependency

Actually, it really is a failing dependency problem. I got this problem because I removed one of my jars on the `.m2` folder. The issue is that Eclipse uses its target folder to compile, but maybe the jar is missing on the local repository. The JavaDoc of `NoClassDefFoundError` is the following:

> Thrown if the Java Virtual Machine or a ClassLoader instance tries to load in the definition of a class (as part of a normal method call or as part of creating a new instance using the new expression)and no definition of the class could be found. The searched-for class definition existed when the currently executing class was compiled, but the definition can no longer be found

Therefore, I would still leave the message.

hohwille commented:
As github prevents me to commont on the old conversation:

> > getType("none.existing.Type"). So LOG.debug("Could not find class {}.", qualifiedName); would make more sense.

> I agree, I have changed it.

Thanks. Great.

> > but must not be a failing dependency

> Actually, it really is a failing dependency problem. I got this problem because I removed one of my jars on the `.m2` folder. The issue is that Eclipse uses its target folder to compile, but maybe the jar is missing on the local repository. The JavaDoc of `NoClassDefFoundError` is the following:

I am not saying that it can be a missing dependency and that me be the reason in your case.
What I wanted to outline with "none.existing.Type" that the Classname my just contain a typo and simply exists nowhere - even if you add all the dependencies of the world.

> > Thrown if the Java Virtual Machine or a ClassLoader instance tries to load in the definition of a class (as part of a normal method call or as part of creating a new instance using the new expression)and no definition of the class could be found. The searched-for class definition existed when the currently executing class was compiled, but the definition can no longer be found

> Therefore, I would still leave the message.

Therefore I still think the logmessage can be wrong and is therefore confusing. All we know is that a class was requested that was not found. We have no clue what the reason for this problem actually is.

hohwille commented:
> I agree that it would be very nice to have a lazily loading of the dependencies. However, this seems a very complex topic. For instance:
> When I try to load EmployeeEntity, I first need to load Employee defined on the api. This is easy because the core already defined api as a dependency. But, Employee extends ApplicationEntity which extends GenericEntity. That class is tricky to load because it is not defined inside the api.
> The api defines as dependency devon4j-rest which defines as dependency devon4j-basic. So it is indirectly referenced.
> If Maven was created, was in order to fix this JAR hell problem. If you run the command mvn dependency:build-classpath in any of your Maven projects, you will see all the jars that are loaded in the class path.
> If you guide me on this, we can try to find an optimal solution. However, my proposed solution already works pretty well. We have even tested using My Thai Star, which as you know, is a large project.

First of all, I do not expect you to solve this in this PR. You already did a great job and this PR is almost ready to be merged.
Second, I fully agree that lazy resolving of dependencies adds a huge complexity and we should start with your current approache that works fine. I can create an issue for the backlog to optimize with lazy loading.

My plan to do lazy loading would be to have a hierarchical classloader that first only resolves the project dependencies itself. As long as only classes and source-code from the main project itself are requested this would work and it would be faster to boostrap. Whenever we hit a class we can not find, we would do a next iteration of resolving dependencies. There might even be some heuristics to check the groupId in relation to the package of the requested class in order to find the dependency and according URL to load from in case of a happy path. Of course complex and tricky. Also I am aware that the first class requested that does not exist at all would force us to resolve all dependencies recursively anyhow. Still I think there might be scenarios where only few things will be requested from `mmm-code` and in such cases the optimization would be beneficial. We could analyze some CobiGen scenarios. If we come to the point that in the end we always need most up to all dependencies resolved, we could handle the optimization with a priority close to infinity ;)

jdiazgon commented:
Thanks for the clarification :)

When I finish all the changes I will inform you

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/JavaContext.java:
```diff
@@ -222,9 +228,34 @@ public BaseGenericType getType(Class<?> clazz) {
+      JavaSourceLoader javaSourceLoader = (JavaSourceLoader) source.getLoader();
+
+      // HERE
+      if (codeSource != source.getReflectiveObject()) {
```
I'm currently debugging this and the problem seems to be when trying to create a new `BaseSource` from a dependency Jar. In that case, `reflectiveObject` is null. 

I'm trying to figure out how to fix this.

jdiazgon commented:
I have implemented all the changes. Please review them when you are free. 

The ones pending are some nice to have and the wrapping lines formatting. Should I change it to your format?

jdiazgon commented:
Hi @hohwille. it would be really interesting for me if we release to Maven the first version of this library  :)

hohwille commented:
> why did you remove static here? This method was meant to be the entry point for a user of the library to create the context easily. Now it would be required that the provider first needs to be created. However that instance does not even to be used in your private method variant createFromLocalMavenProject. IMHO this does not make sense. Why dont you just keep the static here and add static to your new private method? The need to create a provider instance that is then not used but creating a new provider within createFromLocalMavenProject seems odd.

> This issue is still open (neither resolved nor explained). Can you give feedback or fix this please?

I do not see that this was addressed.

You do not need to care about the rest of the formatting and nice-to-haves. But this static "API" changes are still in question for me.

I am fine to create a release and publish to maven central when we have this merged. But I wont be able to do it this week. Next week should be possible.

jdiazgon commented:
> Could you have a look at those comments that are still unresolved?

Now it has been addressed. In order to make those methods again `static` I had to make some other changes, but I think they are fine.

I have also fixed an annoying problem on the pom, in which Eclipse was claiming that there was an error but it is not true. Please take a look [here](https://stackoverflow.com/questions/6352208/how-to-solve-plugin-execution-not-covered-by-lifecycle-configuration-for-sprin) to know what I'm talking about. I just had to add a new tag `<pluginManagement>`

jdiazgon commented:
Wops, there is a test failing. I'm going to check what is going on.

jdiazgon commented:
> I have also fixed an annoying problem on the pom

It seems that my fix broke the test because the exec plugin for building the test project was not being executed. Therefore I have reverted it.

hohwille commented:
> I have also fixed an annoying problem on the pom, in which Eclipse was claiming that there was an error but it is not true. Please take a look here to know what I'm talking about. I just had to add a new tag <pluginManagement>

We do solve this problem in our ide rather than polluting all our POMs with such non-sense. The devon-ide solution is here:
https://github.com/hohwille/devon-ide/blob/master/settings/src/main/settings/eclipse/lifecycle-mapping-metadata.xml
If projects need additional maven plugins they can add this to their forked settings and also provide PR for devon-ide.
Integration for eclipse is configured here:
https://github.com/hohwille/devon-ide/blob/master/settings/src/main/settings/eclipse/workspace/update/.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.m2e.core.prefs#L7

I have no clue if cobigen did not use this feature or whatever. However, it may just be a lack of documentation and knowledge...

hohwille commented:
Would you be able to fix the bug from my review comments which is still unresolved?
look for `getType(e.getMessage())` also for my comment `This is a contextual method call.`
I was hoping you would go through all remaining review comments and either fix that or add a comment with your PoV.

jdiazgon commented:
> I was hoping you would go through all remaining review comments and either fix that or add a comment with your PoV

I think there is some kind of bug in Github because I don't see those review comments that you talk about. For instance:

![image](https://user-images.githubusercontent.com/31688036/60079744-d509f480-972e-11e9-8d98-bbf9d77ce714.png)

I see no review message on that. Also, I have been checking all your review comments I see and either they are resolved, or either they are nice to have.

I will fix this, but if there are still other review comments pending, can you post them again?

hohwille commented:
> I think there is some kind of bug in Github because I don't see those review comments that you talk about.

I am very sorry. It seesm the bug was me. I did provide the review and saw it on github (even after reboot of OS and restart of browser) but it was not "committed" and therefore not visible to others. This obviously explains the problem. Sorry, all my fault...

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java:
```diff
@@ -129,71 +140,276 @@ private BaseSource createSource(Dependency dependency, File byteCodeArtifact) {
    * @return the {@link JavaContext} for the local Maven project in the current working directory.
    */
-  public static JavaContext createFromLocalMavenProject() {
+  public JavaContext createFromLocalMavenProject() {
```
I have already reverted it back to `static` in my latest commits.

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/parser/JavaSourceCodeReaderHighlevel.java:
```diff
@@ -96,21 +96,52 @@ public BaseType parse(Reader reader, BaseFile javaFile) {
+   * @param javaFile the {@link BaseFile} to read.
+   * @return the full qualified name of the Java file.
+   */
+  public String parseQualifiedName(Reader reader, BaseFile javaFile) {
```
I really need this feature for the CLI. The use case is that on the CLI I need to load compiled classes which I only know the name of the class, but I don't know the package. The ``loadClass(...)` method expects full qualified names.

I tried to search for other solutions on internet, but I didn't find anything which could return easily the full qualified name. In the `mmm-code` library seemed to make sense and it was really easy to implement.

If you don't like it there, please provide me an alternative.

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/loader/JavaSourceLoader.java:
```diff
@@ -134,7 +142,7 @@ public BaseGenericType getType(Class<?> clazz) {
   }
 
-  private BaseType getTypeInternal(Class<?> clazz, BasePackage pkg) {
+  public BaseType getTypeInternal(Class<?> clazz, BasePackage pkg) {
```
I can't remember now why did I change it. Anyway, it seems not to be needed, so we can change it back to `private`.

jdiazgon commented on java/impl/src/main/java/net/sf/mmm/code/impl/java/JavaExtendedContext.java:
```diff
@@ -37,27 +68,42 @@ public JavaExtendedContext(BaseSourceImpl source, BaseSourceProvider sourceProvi
 
   @Override
-  protected BaseType getTypeFromCache(String qualifiedName) {
+  public BaseType getRootExceptionType() {
```
Most probably I was doing testing and I forgot to remove it. Doesn't seem to be useful.

jdiazgon commented:
Malte has informed me that you are going to make the release very soon. I'm sorry that I did not have time yet to resolve the last comments, but I am currently busy with the CLI.

I gave already my point of view on some of the comments. The only critical thing for me is to have a method to retrieve the full qualified name. I'm already using it on the CLI and it is working perfectly.

hohwille commented:
I finally decided to take this over as otherwise progress is too slow.
I created PR #19 as a replacement and do the rework there myself.
There are various improvements to do - e.g. in your test the effective model of the core POM is evaluated 3 times rather than just once. This is waste that we need to optimize.

hohwille commented:
All solved, merged and released.
